# Data Preparation

### Contents:
1. Import libraries
2. Import data
3. Data exploration and cleaning
4. Deriving new columns
5. Dropping and renaming columns
6. Checking and changing data types
7. Export data

## 1. Import libraries

In [1]:
# Import the necessary libraries
import pandas as pd
import numpy as np
import os
import datetime as dt

## 2. Import data

In [2]:
# Create a string of the path for the main project folder
path = r'C:\Users\Ryan\Documents\08-31-2023 Citi Bike Analysis'

In [3]:
# Import the “201903-citibike-tripdata.csv” data set using the os library
df_citibike = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', '201903-citibike-tripdata.csv'))

## 3. Data exploration and cleaning

In [4]:
# Check the output
df_citibike.head()

tripduration                 starttime                  stoptime  \
0          1463  2019-03-01 00:00:16.0970  2019-03-01 00:24:39.3880   
1           285  2019-03-01 00:00:32.3850  2019-03-01 00:05:18.1830   
2           686  2019-03-01 00:00:47.7970  2019-03-01 00:12:14.3090   
3           442  2019-03-01 00:01:01.2090  2019-03-01 00:08:23.7510   
4          2913  2019-03-01 00:01:09.2810  2019-03-01 00:49:42.3810   

   start station id    start station name  start station latitude  \
0             319.0  Fulton St & Broadway               40.711066   
1             439.0        E 4 St & 2 Ave               40.726281   
2             526.0       E 33 St & 5 Ave               40.747659   
3            3474.0     6 Ave & Spring St               40.725256   
4             379.0       W 31 St & 7 Ave               40.749156   

   start station longitude  end station id             end station name  \
0               -74.009447           347.0  Greenwich St & W Houston St   
1               -73.989780           150.0            E 2 St & Avenue C   
2               -73.984907          3474.0            6 Ave & Spring St   
3               -74.004121           355.0        Bayard St & Baxter St   
4               -73.991600           212.0      W 16 St & The High Line   

   end station latitude  end station longitude  bikeid    usertype  \
0             40.728846             -74.008591   35618  Subscriber   
1             40.720874             -73.980858   31113  Subscriber   
2             40.725256             -74.004121   19617  Subscriber   
3             40.716021             -73.999744   27086  Subscriber   
4             40.743349             -74.006818   34791  Subscriber   

   birth year  gender  
0        1989       1  
1        1980       1  
2        1987       1  
3        1987       2  
4        1991       1

In [5]:
# Check the dimensions
df_citibike.shape

(1327960, 15)

#### Missing values check:

In [6]:
# Check for null/missing values
df_citibike.isna().sum()

tripduration                0
starttime                   0
stoptime                    0
start station id           10
start station name         10
start station latitude      0
start station longitude     0
end station id             10
end station name           10
end station latitude        0
end station longitude       0
bikeid                      0
usertype                    0
birth year                  0
gender                      0
dtype: int64

In [7]:
# Drop null/missing values
df_citibike = df_citibike.dropna()

In [8]:
# Check the dimensions
df_citibike.shape

(1327950, 15)

#### Duplicates check:

In [9]:
# Check for duplicates
df_citibike[df_citibike.duplicated()].shape

(0, 15)

No duplicates (:

#### Accuracy check:

In [10]:
# Check relevant columns for accuracy
df_citibike[['tripduration','birth year', 'gender']].describe()

tripduration    birth year        gender
count  1.327950e+06  1.327950e+06  1.327950e+06
mean   8.826540e+02  1.979394e+03  1.167328e+00
std    8.571514e+03  1.220167e+01  4.971219e-01
min    6.100000e+01  1.857000e+03  0.000000e+00
25%    3.210000e+02  1.969000e+03  1.000000e+00
50%    5.390000e+02  1.982000e+03  1.000000e+00
75%    9.440000e+02  1.989000e+03  1.000000e+00
max    2.969781e+06  2.003000e+03  2.000000e+00

#### Gender check:

In [11]:
# Obtain frequencies for 'gender' column
df_citibike['gender'].value_counts(dropna = False)

1    962593
2    293780
0     71577
Name: gender, dtype: int64

There are 71,577 unknown genders. Since there is no way to correct these values, they will be removed.

In [12]:
# Remove rows with unknown gender
df_citibike = df_citibike.loc[df_citibike['gender'].isin([1,2])]

In [13]:
# Check the dimensions
df_citibike.shape

(1256373, 15)

In [14]:
df_citibike['gender'] = df_citibike['gender'].replace({1: 'Male', 2: 'Female'})

In [15]:
# Check the out
df_citibike.head()

tripduration                 starttime                  stoptime  \
0          1463  2019-03-01 00:00:16.0970  2019-03-01 00:24:39.3880   
1           285  2019-03-01 00:00:32.3850  2019-03-01 00:05:18.1830   
2           686  2019-03-01 00:00:47.7970  2019-03-01 00:12:14.3090   
3           442  2019-03-01 00:01:01.2090  2019-03-01 00:08:23.7510   
4          2913  2019-03-01 00:01:09.2810  2019-03-01 00:49:42.3810   

   start station id    start station name  start station latitude  \
0             319.0  Fulton St & Broadway               40.711066   
1             439.0        E 4 St & 2 Ave               40.726281   
2             526.0       E 33 St & 5 Ave               40.747659   
3            3474.0     6 Ave & Spring St               40.725256   
4             379.0       W 31 St & 7 Ave               40.749156   

   start station longitude  end station id             end station name  \
0               -74.009447           347.0  Greenwich St & W Houston St   
1               -73.989780           150.0            E 2 St & Avenue C   
2               -73.984907          3474.0            6 Ave & Spring St   
3               -74.004121           355.0        Bayard St & Baxter St   
4               -73.991600           212.0      W 16 St & The High Line   

   end station latitude  end station longitude  bikeid    usertype  \
0             40.728846             -74.008591   35618  Subscriber   
1             40.720874             -73.980858   31113  Subscriber   
2             40.725256             -74.004121   19617  Subscriber   
3             40.716021             -73.999744   27086  Subscriber   
4             40.743349             -74.006818   34791  Subscriber   

   birth year  gender  
0        1989    Male  
1        1980    Male  
2        1987    Male  
3        1987  Female  
4        1991    Male

#### Birth year check:

The minimum value for the birth year is 1857, implying the oldest Citi Bike user is 162 years old! For analysis, we will remove any Citi Bike user older than 85 years.

In [16]:
# Check how many records have birth year < 1934
df_citibike.loc[df_citibike['birth year'] < 1934].shape

(881, 15)

In [17]:
# Remove rows with birth year < 1934
df_citibike = df_citibike.loc[df_citibike['birth year'] >= 1934]

In [18]:
# Check the dimensions
df_citibike.shape

(1255492, 15)

#### Trip duration check:

The maximum value for the trip duration is 2,969,781 seconds. This is an usually large trip duration, but it is plausible that Citi Bike users may borrow a bike for long periods of time, especially if they are subscribers. We will further explore the trip duration to see how many records have long trip duration.

In [19]:
# Determine the upper outlier boundary for trip duration
944 + 1.5*(944-321)

1878.5

In [20]:
# Check how many rows have a trip duration > 1878.5
df_citibike.loc[df_citibike['tripduration'] > 1878.5].shape

(66070, 15)

There are 66,070 outlier values in the tripduration column. Removing these outliers would result in removing more than 5% of the data set, which could cause major issues in our analysis. Therefore, we will leave these extreme values.

In [21]:
# Re-check relevant columns for accuracy
df_citibike[['tripduration','birth year']].describe()

tripduration    birth year
count  1.255492e+06  1.255492e+06
mean   8.158086e+02  1.979952e+03
std    7.509738e+03  1.208571e+01
min    6.100000e+01  1.934000e+03
25%    3.150000e+02  1.971000e+03
50%    5.240000e+02  1.983000e+03
75%    9.020000e+02  1.990000e+03
max    2.969781e+06  2.003000e+03

## 4. Deriving new columns

In [22]:
# Convert trip duration to minutes
df_citibike['trip_duration'] = round(df_citibike['tripduration']/60, 2)

In [23]:
# Create age column
df_citibike['age'] = 2019 - df_citibike['birth year']

In [24]:
# Create age_group column
df_citibike.loc[df_citibike['age'] <= 25, 'age_group'] = '16-25'
df_citibike.loc[(df_citibike['age'] >= 26) & (df_citibike['age'] <= 35), 'age_group'] = '26-35'
df_citibike.loc[(df_citibike['age'] >= 36) & (df_citibike['age'] <= 45), 'age_group'] = '36-45'
df_citibike.loc[(df_citibike['age'] >= 46) & (df_citibike['age'] <= 55), 'age_group'] = '46-55'
df_citibike.loc[(df_citibike['age'] >= 56) & (df_citibike['age'] <= 65), 'age_group'] = '56-65'
df_citibike.loc[(df_citibike['age'] >= 66) & (df_citibike['age'] <= 75), 'age_group'] = '66-75'
df_citibike.loc[df_citibike['age'] >= 76, 'age_group'] = '76-85'

In [25]:
# Convert 'starttime' and 'stoptime' columns to pandas datetime
df_citibike['starttime'] = pd.to_datetime(df_citibike['starttime'], format="%Y-%m-%d %H:%M:%S")
df_citibike['stoptime'] = pd.to_datetime(df_citibike['stoptime'], format="%Y-%m-%d %H:%M:%S")

In [26]:
# Create start_hour column
df_citibike['start_hour'] = df_citibike['starttime'].dt.hour

In [27]:
# Create a start_day column
df_citibike['start_day'] = df_citibike['starttime'].dt.day

In [28]:
# Create day_of_week column
df_citibike['start_day_name'] = df_citibike['starttime'].dt.day_name()

## 5. Dropping and renaming columns

In [29]:
# Drop columns not relevant for analysis
df_citibike = df_citibike.drop(columns = ['tripduration', 'start station id', 'end station id', 'bikeid', 'birth year', 'starttime', 'stoptime'])

In [30]:
# Check the columns
df_citibike.columns

Index(['start station name', 'start station latitude',
       'start station longitude', 'end station name', 'end station latitude',
       'end station longitude', 'usertype', 'gender', 'trip_duration', 'age',
       'age_group', 'start_hour', 'start_day', 'start_day_name'],
      dtype='object')

In [31]:
# Rename columns to follow proper naming conventions
df_citibike.rename(columns = {'start station name': 'start_station',
                              'start station latitude': 'start_latitude',
                              'start station longitude': 'start_longitude',
                              'end station name': 'end_station',
                              'end station latitude': 'end_latitude',
                              'end station longitude': 'end_longitude',
                              'usertype': 'user_type'},
                   inplace = True)

In [32]:
# Rearrange columns
df_citibike = df_citibike[[
    'user_type',
    'gender',
    'age',
    'age_group',
    'trip_duration',
    'start_day_name',
    'start_day',
    'start_hour',
    'start_station',
    'start_latitude',
    'start_longitude',
    'end_station',
    'end_latitude',
    'end_longitude'
]]

In [33]:
# Check the output
df_citibike.head()

user_type  gender  age age_group  trip_duration start_day_name  start_day  \
0  Subscriber    Male   30     26-35          24.38         Friday          1   
1  Subscriber    Male   39     36-45           4.75         Friday          1   
2  Subscriber    Male   32     26-35          11.43         Friday          1   
3  Subscriber  Female   32     26-35           7.37         Friday          1   
4  Subscriber    Male   28     26-35          48.55         Friday          1   

   start_hour         start_station  start_latitude  start_longitude  \
0           0  Fulton St & Broadway       40.711066       -74.009447   
1           0        E 4 St & 2 Ave       40.726281       -73.989780   
2           0       E 33 St & 5 Ave       40.747659       -73.984907   
3           0     6 Ave & Spring St       40.725256       -74.004121   
4           0       W 31 St & 7 Ave       40.749156       -73.991600   

                   end_station  end_latitude  end_longitude  
0  Greenwich St & W Houston St     40.728846     -74.008591  
1            E 2 St & Avenue C     40.720874     -73.980858  
2            6 Ave & Spring St     40.725256     -74.004121  
3        Bayard St & Baxter St     40.716021     -73.999744  
4      W 16 St & The High Line     40.743349     -74.006818

In [34]:
# Check the dimensions
df_citibike.shape

(1255492, 14)

## 6. Checking and changing data types

In [35]:
# Check for mixed-type data
for col in df_citibike.columns.tolist():
    weird = (df_citibike[[col]].applymap(type) != df_citibike[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_citibike[weird]) > 0:
        print (col)

No mixed-type data

In [36]:
# Check the data types
df_citibike.dtypes

user_type           object
gender              object
age                  int64
age_group           object
trip_duration      float64
start_day_name      object
start_day            int64
start_hour           int64
start_station       object
start_latitude     float64
start_longitude    float64
end_station         object
end_latitude       float64
end_longitude      float64
dtype: object

In [37]:
# Change data types to more optimal ones
df_citibike['age'] = df_citibike['age'].astype('uint8')
df_citibike['trip_duration'] = df_citibike['trip_duration'].astype('float32')
df_citibike['start_day'] = df_citibike['start_day'].astype('uint8')
df_citibike['start_hour'] = df_citibike['start_hour'].astype('uint8')
df_citibike['start_latitude'] = df_citibike['start_latitude'].astype('float32')
df_citibike['start_longitude'] = df_citibike['start_longitude'].astype('float32')
df_citibike['end_latitude'] = df_citibike['end_latitude'].astype('float32')
df_citibike['end_longitude'] = df_citibike['end_longitude'].astype('float32')

In [38]:
# Re-check relevant columns for accuracy
df_citibike[['age', 'trip_duration', 'start_day', 'start_hour']].describe()

age  trip_duration     start_day    start_hour
count  1.255492e+06   1.255492e+06  1.255492e+06  1.255492e+06
mean   3.904832e+01   1.359681e+01  1.750158e+01  1.388825e+01
std    1.208571e+01   1.249650e+02  8.560315e+00  4.754976e+00
min    1.600000e+01   1.020000e+00  1.000000e+00  0.000000e+00
25%    2.900000e+01   5.250000e+00  1.100000e+01  1.000000e+01
50%    3.600000e+01   8.730000e+00  1.800000e+01  1.500000e+01
75%    4.800000e+01   1.503000e+01  2.500000e+01  1.800000e+01
max    8.500000e+01   4.949635e+04  3.100000e+01  2.300000e+01

## 7. Export data

In [39]:
# Export dataframe
df_citibike.to_pickle(os.path.join(path, '02 Data','Prepared Data', '201903-citibike-tripdata-cleaned.pkl'))